### PR TITLE
Some interface changes (improvements?)

### PR DESCRIPTION
**Pull request by mloughran**
I've made quite a few changes in my fork of em-zeromq - various refactorings, code simplifications, and changes to the interface which imho make the library easier to use. I wrote them some time ago for my own use, so they may be slightly trigger happy, but I wanted to make you aware of the fork in case the changes are of interest. They're stood up fairly well to our use over the past few months.

Obviously tests are lacking: I'm happy to spend some time in that department if you wanted to accept the changes.

Cheers,
Martyn

### CODE REVIEW

File: lib/em-zeromq/socket.rb [hunk 1]
```diff
@@ -124,25 +114,9 @@ def notify_readable
...
-              end
-            end
-            
-            @handler.on_readable(self, msg_parts)
-          else
-            break
-          end
+
+        while (message = get_message)
+          emit(:message, *message)
```
Comment (schmurfy) [inline lib/em-zeromq/socket.rb:119]: Where and how are handled multi-part messages ?
Unless I miss something this only works correctly for one part messages

**Comment by schmurfy** [comment]
Overall it looks fine except for my comment above.
I am not entirely convinced by the event emitter but it does not bothers me that much.

There are some changes you made to the context/socket class I wanted to add (and in fact are but sleeping on a local branch xD).

**Comment by schmurfy** [comment]
@mloughran can you answer the line comment I added about multi-part ?
Once this is solved I have no problem merging the pull request.

**Comment by skull-squadron** [comment]
+1

File: lib/em-zeromq/socket.rb (hunk 1 above)
Comment (mloughran) [inline lib/em-zeromq/socket.rb:119]: Sorry for disappearing from this discussion. Multi part does work fine - `get_message` returns an array of parts which are then splatted, so the interface for receiving is:

```
pull.on(:message) { |part1, part2|
  p [:part1, part1.copy_out_string]
}
```

Or if you don't know how many parts to expect:

```
pull.on(:message) { |*parts|
  p [:parts, parts.map(&:copy_out_string)]
}
```

**Comment by mloughran** [comment]
Thanks Julien, I'll take a look at fixing the specs now

**Comment by mloughran** [comment]
@schmurfy the tests have now been updated and are all passing. Let me know if there's anything else you need.

I'm guessing you'll want to release this as 0.4 since it breaks a few interfaces.

**Comment by schmurfy** [comment]
I will try to have a look at this during the week end, now that eventmachine 1.0.0 is released we can switch to it too.

**Comment by schmurfy** [comment]
Since other people complained about the current interface I guess we can give this new syntax a try, if you can just update the example in the README I will merge it.

**Comment by mloughran** [comment]
@schmurfy I've updated the readme and example. Also added some other useful examples in case they're useful to anybody. What do you think about making the example in the readme simpler - like `simpler.rb`?

**Comment by andrewvc** [comment]
Fantastic, looks good to me. @schmurfy your thoughts?

**Comment by schmurfy** [comment]
no problem for me I just did not found the time to correctly review the patch before merging it, I have no problem with changing the example in the README either, go for it :)

**Comment by mloughran** [comment]
@schmurfy have updated readme

**Comment by andrewvc** [comment]
I'll check it out one very last time more fully today or tomorrow and I'll merge it. I'm not anticipating any changes, just want to make sure the examples run and no obvious bugs.

**Comment by andrewvc** [comment]
FYI, 0.4.0 has been released including:
1. These changes
2. eventmachine dep bumped to 1.0.0

**Comment by mloughran** [comment]
Thanks guys

**Pull request merged**